src/cpp/session/modules/SessionThemes.cpp

```diff
@@ -82,69 +128,112 @@ void getThemesInLocation(const rstudio::core::FilePath& location, ThemeMap& them
+               isDark = strIsTrue(matches[1]);
             }
             if ((matches.size() < 2) ||
                 (!isDark &&
-                !std::regex_match(
-                   std::string(matches[1]),
-                   std::regex("(?:0|f(?:alse)?))", std::regex_constants::icase))))
+                !strIsFalse(matches[1])))
             {
-               // TODO: warning / logging about using default isDark value.
+               LOG_WARNING_MESSAGE("rs-theme-is-dark is not set or not convertable to boolean for theme \"" + name + "\".");
             }
 
-            themeMap[name] = std::pair<std::string, bool>(k_themeFileStr, isDark);
+            themeMap[boost::algorithm::to_lower_copy(name)] = std::make_tuple(
+               name,
+               urlPrefix + themeFile.filename(),
+               isDark);
          }
       }
    }
 }
 
 /**
- * @brief Gets the list of all RStudio editor themes.
+ * @brief Gets the location of themes that are installed with RStudio.
  *
- * @return The list of all RStudio editor themes.
+ * @return The location of themes that are installed with RStudio.
  */
-SEXP rs_getThemes()
+FilePath getDefaultThemePath()
+{
+   return session::options().rResourcesPath().childPath("themes");
+}
+
+/**
+ * @brief Gets the location of custom themes that are installed for all users.
+ *
+ * @return The location of custom themes that are installed for all users.
+ */
+FilePath getGlobalCustomThemePath()
 {
-   // List all files in the global location first.
-   // TODO: Windows
    using rstudio::core::FilePath;
 
+   const char* kGlobalPathAlt = std::getenv("RS_THEME_GLOBAL_HOME");
+   if (kGlobalPathAlt)
+   {
+      return FilePath(kGlobalPathAlt);
+   }
+
 #ifdef _WIN32
-   FilePath globalPath = rstudio::core::system::systemSettingsPath("RStudio\\themes", false);
-   FilePath localPath = rstudio::core::system::userHomePath().childPath(".R\\rstudio\\themes");
+   return core::system::systemSettingsPath("RStudio\\themes", false);
 #else
-   FilePath globalPath("/etc/rstudio/themes/");
-   FilePath localPath = rstudio::core::system::userHomePath().childPath(".R/rstudio/themes/");
+   return FilePath("/etc/rstudio/themes/");
 #endif
+}
 
-   const char* k_globalPathAlt = std::getenv("RS_THEME_GLOBAL_HOME");
-   const char* k_localPathAlt = std::getenv("RS_THEME_LOCAL_HOME");
-   if (k_globalPathAlt)
-   {
-      globalPath = FilePath(k_globalPathAlt);
-   }
-   if (k_localPathAlt)
+/**
+ * @brief Gets the location of custom themes that are installed for the current user.
+ *
+ * @return The location of custom themes that are installed for the current user.
+ */
+FilePath getLocalCustomThemePath()
+{
+   using rstudio::core::FilePath;
+   const char* kLocalPathAlt = std::getenv("RS_THEME_LOCAL_HOME");
+   if (kLocalPathAlt)
    {
-      localPath = FilePath(k_localPathAlt);
+      return FilePath(kLocalPathAlt);
    }
 
+#ifdef _WIN32
+   return core::system::userHomePath().childPath(".R\\rstudio\\themes");
```

Comment:
This is rather non-Windows-ish, and ends up in `Documents\.R`,  and isn't a hidden directory, so users are going to see both .R and R. Looks like we put `keybindings` there, too, and maybe other things (didn't look too hard). Probably not worth changing at this point, so just FYI.

![2018-06-28_16-25-20](https://user-images.githubusercontent.com/10569626/42065519-22e91142-7af0-11e8-91f3-c621390e23a6.png)

Reply:
It is a bit odd that there's both `.R` and `R`, however the intention was for the custom themes to be saved alongside the keybindings preferences.